cli/tests/test_help_command.rs

```diff
@@ -0,0 +1,46 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::common::TestEnvironment;
+
+#[test]
+fn test_help() {
+    let test_env = TestEnvironment::default();
+
+    let help_cmd_stdout = test_env.jj_cmd_success(test_env.env_root(), &["help"]);
+    // The help command output should be equal to the long --help flag
+    let help_flag_stdout = test_env.jj_cmd_success(test_env.env_root(), &["--help"]);
+    assert_eq!(help_cmd_stdout, help_flag_stdout);
+
+    // Help command should work with commands
+    let help_cmd_stdout = test_env.jj_cmd_success(test_env.env_root(), &["help", "log"]);
+    let help_flag_stdout = test_env.jj_cmd_success(test_env.env_root(), &["log", "--help"]);
+    assert_eq!(help_cmd_stdout, help_flag_stdout);
+
+    // Help command should work with subcommands
+    let help_cmd_stdout =
+        test_env.jj_cmd_success(test_env.env_root(), &["help", "workspace", "add"]);
+    let help_flag_stdout =
+        test_env.jj_cmd_success(test_env.env_root(), &["workspace", "add", "--help"]);
+    assert_eq!(help_cmd_stdout, help_flag_stdout);
```

Comment:
We should have a test for e.g. `jj new help main` (which is supposed to create a merge commit on top of the `help` and `main` commits. That doesn't work with the current version of this PR, so the PR needs to be fixed. Maybe the best way of fixing it is revert to the version with the real `help` command?

Reply:
Oh, maybe we'll have to set `.disable_help_subcommand(true)`, etc. to all subcommands recursively? That's a bit more than I thought.

BTW, if we take this approach, we can enable the hack only after clap returned a `DisplayHelp` error.

Reply:
> BTW, if we take this approach, we can enable the hack only after clap returned a `DisplayHelp` error.

The issue is that when `--no-pager` is used after a `help` command, a `DisplayHelp` is not returned.

Reply:
> We should have a test for e.g. `jj new help main` (which is supposed to create a merge commit on top of the `help` and `main` commits). That doesn't work with the current version of this PR, so the PR needs to be fixed. Maybe the best way of fixing it is revert to the version with the real `help` command?

I think maybe there are ways to fix that in the current version, but I also think that new bugs like this might appear. So I think the version with a real `help` command is a more solid choice.

Reply:
> The issue is that when `--no-pager` is used after a `help` command, a `DisplayHelp` is not returned.

Got it. It's unfortunate that `help` doesn't ignore unknown args as `--help` would do.

> I think maybe there are ways to fix that in the current version, but I also think that new bugs like this might appear. So I think the version with a real `help` command is a more solid choice.

Suppose we have to modify all subcommands recursively anyway, I agree it might be simpler to just duplicate `help` commands.

Reply:
Having our own `help` command will also be useful for implementing the `jj help <category>` if we ever want something like that. The `<category>` there might be e.g. `revset`.

Oh, because I figured I should explain what I meant by `<category>` I also happened remind myself of the other type of help categories we want, which is that the top-level `jj help` would be easier to read if we grouped commands into categories.

Reply:
I agree we'll eventually need our own `jj help`. I don't think `jj <subcommand> help` should support all the arguments (such as `jj <subcommand> help revset`), though.

Reply:
> Having our own `help` command will also be useful for implementing the `jj help <category>` if we ever want something like that. The `<category>` there might be e.g. `revset`.
> 
> Oh, because I figured I should explain what I meant by `<category>` I also happened remind myself of the other type of help categories we want, which is that the top-level `jj help` would be easier to read if we grouped commands into categories.

Interesting, I like the idea to not have to go to the docs for information on these sorts of things. I might try to work on something along those lines after this is accepted.

Reply:
> I don't think `jj <subcommand> help` should support all the arguments (such as `jj <subcommand> help revset`), though.

I agree. I was surprised to see that that currently works. Dropping that support seems like a relatively minor breaking change.

> Interesting, I like the idea to not have to go to the docs for information on these sorts of things. I might try to work on something along those lines after this is accepted.

Thanks! Just keep in mind that we want a way of not having to duplicate documentation. We've talked about this somewhere in the past but I'm not sure if it was on GitHub or on Discord. I'd rather just point to the website than to duplicate content, but it would be nice if we can find a way to share the content.

Reply:
> > I don't think `jj <subcommand> help` should support all the arguments (such as `jj <subcommand> help revset`), though.
> 
> I agree. I was surprised to see that that currently works. Dropping that support seems like a relatively minor breaking change.

Do you mean we can deprecate/remove `jj <subcommand> help` in favor of `jj help <subcommand>`? I agree that as we can remove hacks from this PR.

Reply:
We would deprecate `jj <subcommand> help` but keep `jj <subcommand> help <subsubcommand>`? Or we would just have a primary `help` command and no sub ones?

Reply:
Basically only letting the user get help for a sub-subcommand with `jj help <subcommand> <sub-subcommand>`

Reply:
Ohhhh, that makes sense. I like this, it reduces code complexity, too ;).